iphone/cli/lib/process-js-task.js

```diff
@@ -0,0 +1,374 @@
+		const unchangedFiles = Array.from(this.inputFiles).filter(filePath => !changedFiles.has(filePath));
+		const unchangedPromise = Promise.all(unchangedFiles.map(filePath => this.processJsFile(filePath)));
+
+		return Promise.all([ deletedPromise, updatedPromise, unchangedPromise ]);
+	}
+
+	/**
+	 * Loads last run's task data if this task will be skipped entirely.
+	 *
+	 * Note that this function will be called when no input files changed. Since
+	 * we also depend on some config values from the builder this is used to
+	 * fallback to a full task run if required.
+	 *
+	 * @return {Promise}
+	 */
+	loadResultAndSkip() {
+		const loaded = this.loadTaskData();
+		const fullBuild = !loaded || this.requiresFullBuild();
+		if (fullBuild) {
+			return this.doFullTaskRun();
+		}
+
+		return Promise.resolve();
+	}
+
+	/**
+	 * Function that will be called after the task run finished.
+	 *
+	 * Used to populate values back into the builder and update the task data
+	 * before it gets written to disk.
+	 */
+	afterTaskAction() {
+		this.builder.tiSymbols = this.data.tiSymbols;
+
+		this.data.jsBootstrapFiles = this.jsBootstrapFiles;
+		this.data.jsFiles = this.jsFiles;
+		this.data.analyzeOptionsHash = this.generateHash(JSON.stringify(this.defaultAnalyzeOptions));
+
+		this.saveTaskData();
+
+		this.builder.unmarkBuildDirFiles(this.incrementalDirectory);
+	}
+
+	/**
+	 * Processes a single JavaScript file by applying several transforms and then copying it to
+	 * its destination path.
+	 *
+	 * If the content of the unprocessed JavaScript file did not change compared to the last run,
+	 * any further processing will be skipped.
+	 *
+	 * @param {String} filePathAndName Full path to the JavaScript file
+	 * @return {Promise}
+	 */
+	processJsFile(filePathAndName) {
+		let file = this.resolveRelativePath(filePathAndName);
+		if (!file) {
+			throw new Error(`Unable to resolve relative path for ${filePathAndName}.`);
+		}
+
+		const bootstrapPath = file.substr(0, file.length - 3);  // Remove the ".js" extension.
+		if (bootstrapPath.endsWith('.bootstrap') && !this.jsBootstrapFiles.includes(bootstrapPath)) {
+			this.jsBootstrapFiles.push(bootstrapPath);
+		}
+
+		const info = this.jsFiles[file];
+		if (this.builder.encryptJS) {
+			if (file.indexOf('/') === 0) {
+				file = path.basename(file);
+			}
+			this.builder.jsFilesEncrypted.push(file); // original name
+			file = file.replace(/\./g, '_');
+			info.dest = path.join(this.builder.buildAssetsDir, file);
+			this.builder.jsFilesToEncrypt.push(file); // encrypted name
+		}
+
+		return new Promise((resolve) => {
+			try {
+				this.builder.cli.createHook('build.ios.copyResource', this.builder, (from, to, cb) => {
+					const originalContents = fs.readFileSync(from).toString();
+
+					const r = {
+						original: originalContents,
+						contents: originalContents,
+						symbols: []
+					};
+
+					this.builder.cli.createHook('build.ios.compileJsFile', this.builder, (r, from, to, cb2) => {
```

Comment:
Also, dumb to never look into this myself, but do we need to create a hook *for every file*? Can't we generate these two hooks once, and just call/pass in different values when we invoke them?